xmtp_id/src/credential_verifier.rs

```diff
@@ -0,0 +1,99 @@
+use prost::Message;
+use xmtp_mls::{
+    credential::{AssociationError, Credential},
+    types::Address,
+};
+use xmtp_proto::xmtp::mls::message_contents::MlsCredential as CredentialProto;
+
+#[derive(Debug, Clone)]
+pub enum AssociationType {
+    ExternallyOwned,
+    #[allow(dead_code)]
+    SmartContract,
+    Legacy,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum VerificationError {
+    #[error(
+        "Address mismatch in Association: Provided:{provided_addr:?} != signed:{signing_addr:?}"
+    )]
+    AddressMismatch {
+        provided_addr: Address,
+        signing_addr: Address,
+    },
+    #[error("Installation public key mismatch")]
+    InstallationPublicKeyMismatch,
+    #[error("protobuf deserialization: {0}")]
+    Deserialization(#[from] prost::DecodeError),
+    #[error("bad association: {0}")]
+    BadAssocation(#[from] AssociationError),
+}
+
+pub struct VerifiedCredential {
+    pub account_address: String,
+    pub account_type: AssociationType,
+}
+
+impl VerifiedCredential {
+    pub fn account_address(&self) -> &String {
+        &self.account_address
+    }
+}
+
+pub struct VerificationRequest {
+    installation_public_key: Vec<u8>,
+    credential: Vec<u8>,
+}
+
+impl VerificationRequest {
+    pub fn new<I, C>(installation_public_key: I, credential: C) -> Self
+    where
+        I: AsRef<[u8]>,
+        C: AsRef<[u8]>,
+    {
+        Self {
+            installation_public_key: installation_public_key.as_ref().to_vec(),
+            credential: credential.as_ref().to_vec(),
+        }
+    }
+}
+
+type VerificationResult = Result<VerifiedCredential, VerificationError>;
+
+#[async_trait::async_trait]
+pub trait CredentialVerifier {
+    async fn verify_credential(request: VerificationRequest) -> VerificationResult;
+    async fn batch_verify_credentials(
+        credentials_to_verify: Vec<VerificationRequest>,
+    ) -> Vec<VerificationResult> {
+        let mut results = Vec::new();
+        for credential in credentials_to_verify {
+            results.push(Self::verify_credential(credential));
+        }
+
+        futures::future::join_all(results).await
+    }
+}
+
+#[async_trait::async_trait]
+impl CredentialVerifier for Credential {
```

Comment:
I feel like it would be better if we could somehow do `impl CredentialVerifier for GrantMessagingAccess` and `impl CredentialVerifier for LegacyCreateIdentity` in order to further decouple legacy verification (v2) and current v3 verification. Then, we could potentially introduce another trait, `Credential` which would replace the enum and maybe clean things up elsewhere as well.

The problem is we don't know the credential type before decoding, making the implementation more complicated and possibly not worth it.

In the interest of not blocking @37ng and moving forward with SCW support going to leave that problem for another day. Eventually V2 will be phased out anyway. The `CredentialVerifier` trait works well with what we have so far

Reply:
I can see how that'd be tricky. We'd need to have two steps of extraction I guess. 

This seems fine for now and we can always refactor further down the line if we think it makes sense.